src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,231 @@
+        {
+            var newSolution = solution;
+
+            foreach (var projectId in solution.ProjectIds)
+            {
+                var project = newSolution.GetProject(projectId);
+                var newProject = await AddMissingImportsAsync(project, cancellationToken).ConfigureAwait(false);
+                newSolution = newProject.Solution;
+            }
+
+            return newSolution;
+        }
+
+        public async Task<Project> AddMissingImportsAsync(Project project, CancellationToken cancellationToken)
+        {
+            var newProject = project;
+
+            foreach (var documentId in project.DocumentIds)
+            {
+                var document = newProject.GetDocument(documentId);
+                newProject = await AddMissingImportsAsync(document, cancellationToken).ConfigureAwait(false);
+            }
+
+            return newProject;
+        }
+
+        public async Task<Project> AddMissingImportsAsync(Document document, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, cancellationToken).ConfigureAwait(false);
+            return await FixDocument(document, diagnostics, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<Project> AddMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            return await FixDocument(document, diagnostics, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<bool> IsMissingImportsAsync(Document document, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, cancellationToken).ConfigureAwait(false);
+            return !diagnostics.IsEmpty;
+        }
+
+        public async Task<bool> IsMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            return !diagnostics.IsEmpty;
+        }
+
+        private Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync(Document document, CancellationToken cancellationToken)
+        {
+            if (!document.TryGetText(out var text))
+            {
+                return Task.FromResult(ImmutableArray<Diagnostic>.Empty);
+            }
+
+            var textSpan = new TextSpan(0, text.Length);
+            return GetDiagnosticsAsync(document, textSpan, cancellationToken);
+        }
+
+        private async Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            if (!document.SupportsSemanticModel)
```

Comment:
it should check both SupportSyntax and Semantic

Reply:
I think this check should go up higher than here. this is too bottom for this kind of check. if there is reason we should put it here, then I think this method might be better to changed to TryGetDia...

Reply:
Sorry, Seems it was leftover from before I switched to using the `DiagnosticAnalyzerService`